frontend/rust-lib/flowy-folder/src/search/folder_indexer.rs

```diff
@@ -0,0 +1,123 @@
+use std::sync::Arc;
+
+use collab::core::collab::{IndexContent, IndexContentReceiver};
+use collab_document::document::DocumentIndexContent;
+use collab_folder::ViewIndexContent;
+use flowy_error::FlowyError;
+use flowy_folder_deps::entities::SearchData;
+use lib_dispatch::prelude::af_spawn;
+use lib_infra::async_trait::async_trait;
+use tracing::error;
+
+/// A trait for folder indexing storage.
+pub trait FolderIndexStorage: Send + Sync {
+  fn search(&self, s: &str, limit: Option<i64>) -> Result<Vec<SearchData>, FlowyError>;
+  fn add_view(&self, id: &str, content: &str) -> Result<(), FlowyError>;
+  fn update_view(&self, id: &str, content: &str) -> Result<(), FlowyError>;
+  fn remove_view(&self, ids: &[String]) -> Result<(), FlowyError>;
+  fn add_document(&self, view_id: &str, page_id: &str, content: &str) -> Result<(), FlowyError>;
+  fn update_document(&self, view_id: &str, page_id: &str, content: &str) -> Result<(), FlowyError>;
+  fn remove_document(&self, page_ids: &[String]) -> Result<(), FlowyError>;
+}
+
+#[async_trait]
+pub trait DocumentIndexContentGetter: Send + Sync {
+  async fn get_document_index_content(
+    &self,
+    doc_id: &str,
+  ) -> Result<DocumentIndexContent, FlowyError>;
+}
+
+pub struct FolderIndexer {
+  storage: Arc<dyn FolderIndexStorage>,
+  document_index_content_getter: Arc<dyn DocumentIndexContentGetter>,
+}
+
+impl FolderIndexer {
+  pub fn new<S, D>(storage: S, doc_getter: D) -> Self
+  where
+    S: FolderIndexStorage + 'static,
+    D: DocumentIndexContentGetter + 'static,
+  {
+    Self {
+      storage: Arc::new(storage),
+      document_index_content_getter: Arc::new(doc_getter),
+    }
+  }
+
+  pub fn search(&self, s: &str, limit: Option<i64>) -> Result<Vec<SearchData>, FlowyError> {
```

Comment:
How do you handle if keep typing that will generate too much request?

Reply:
The client will be implementing debouncing, ensuring that every new character won't trigger a search.

Were you thinking of implementing a rate limit for search?

Reply:
Good point. The following are some ways to work around this:
1. Perform the search only after the user completes typing and presses enter to search. This is the easiest solution with the limitation that there is no search assistance while users are typing.
2. Debounce by waiting for a certain amount of time after the user stops typing before executing the search. This can be done in two places:
	1. Frontend. The frontend contains the state of the search input as the user is typing, and determines when to send the search request to the backend, so it makes sense to perform debouncing here. This is also a common approach for handling inputs.
	2. Backend. Since the backend is not stateful and does not have the state of the search input as the user is typing, it does not make sense to perform debouncing in the backend.
3. Limit dynamic search scope to improve performance where dynamic refers to the period when the user is entering the search term, and only performing the full search against the document content after the user presses enter. Limiting dynamic search to page titles only (as a means of search suggestion), and limiting the number of suggestions to roughly 10.

I propose that a combination of debouncing the user input in the frontend (2i) and limiting dynamic search scope to page titles (3) will be effective in dealing with the performance concerns.

Let me know what you think.